File: app/services/platforms/youtube.py
import json

import httpx

from app.models.platform_account import PlatformAccount
from app.models.post import Post
from app.services.crypto import decrypt
from app.services.platforms.base import BasePlatformClient

YOUTUBE_API = "https://www.googleapis.com/youtube/v3"
YOUTUBE_UPLOAD = "https://www.googleapis.com/upload/youtube/v3/videos"
YOUTUBE_ANALYTICS = "https://youtubeanalytics.googleapis.com/v2/reports"


class YouTubeClient(BasePlatformClient):
    async def post(self, account: PlatformAccount, post: Post) -> str:
        token = await self._get_valid_token(account)

        if post.video_file_path:
            return await self._upload_video(account, post, token)
        else:
            return await self._community_post(account, post, token)

    async def _upload_video(self, account: PlatformAccount, post: Post, token: str) -> str:
        tags = json.loads(post.video_tags or "[]")
        metadata = {
            "snippet": {
                "title": post.video_title or post.content[:100],
                "description": post.content,
                "tags": tags,
                "categoryId": "22",
            },
            "status": {"privacyStatus": "public"},
        }

        # Resumable upload
        async with httpx.AsyncClient(timeout=300.0) as client:
            # Initiate upload
            init_resp = await client.post(
                f"{YOUTUBE_UPLOAD}?uploadType=resumable&part=snippet,status",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                    "X-Upload-Content-Type": "video/*",
                },
                json=metadata,
            )
            init_resp.raise_for_status()
            upload_url = init_resp.headers["Location"]

            # Upload file content
            with open(post.video_file_path, "rb") as f:
                video_data = f.read()

            upload_resp = await client.put(
                upload_url,
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "video/*",
                },
                content=video_data,
            )
            upload_resp.raise_for_status()
            data = upload_resp.json()
            return data["id"]

    async def _community_post(self, account: PlatformAccount, post: Post, token: str) -> str:
        channel_id = account.account_id_on_platform
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{YOUTUBE_API}/commentThreads?part=snippet",
                headers={"Authorization": f"Bearer {token}"},
                json={
                    "snippet": {
                        "channelId": channel_id,
                        "topLevelComment": {
                            "snippet": {"textOriginal": post.content}
                        },
                    }
                },
            )
            resp.raise_for_status()
            return resp.json()["id"]

    async def get_metrics(self, account: PlatformAccount, platform_post_id: str) -> dict:
        token = await self._get_valid_token(account)
        channel_id = account.account_id_on_platform

        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{YOUTUBE_API}/videos",
                headers={"Authorization": f"Bearer {token}"},
                params={
                    "part": "statistics",
                    "id": platform_post_id,
                },
            )
            if resp.status_code != 200:
                return {"likes": 0, "reposts": 0, "replies": 0, "views": 0, "clicks": 0}
            items = resp.json().get("items", [])
            if not items:
                return {"likes": 0, "reposts": 0, "replies": 0, "views": 0, "clicks": 0}

            stats = items[0].get("statistics", {})
            return {
                "likes": int(stats.get("likeCount", 0)),
                "reposts": 0,
                "replies": int(stats.get("commentCount", 0)),
                "views": int(stats.get("viewCount", 0)),
                "clicks": int(stats.get("favoriteCount", 0)),
                "raw_data": json.dumps(stats),
            }

    async def _get_valid_token(self, account: PlatformAccount) -> str:
        from datetime import datetime, timedelta, timezone

        from app.config import settings
        from app.services.crypto import encrypt

        token = decrypt(account.access_token_encrypted)
        if account.token_expires_at:
            now = datetime.now(timezone.utc)
            exp = account.token_expires_at.replace(tzinfo=timezone.utc)
            if exp < now:
                # Refresh
                if account.refresh_token_encrypted:
                    refresh_token = decrypt(account.refresh_token_encrypted)
                    async with httpx.AsyncClient() as client:
                        resp = await client.post(
                            "https://oauth2.googleapis.com/token",
                            data={
                                "grant_type": "refresh_token",
                                "refresh_token": refresh_token,
                                "client_id": settings.YOUTUBE_CLIENT_ID,
                                "client_secret": settings.YOUTUBE_CLIENT_SECRET,
                            },
                        )
                        resp.raise_for_status()
                        data = resp.json()
                        token = data["access_token"]
                        account.access_token_encrypted = encrypt(token)
                        account.token_expires_at = datetime.now(timezone.utc) + timedelta(
                            seconds=data.get("expires_in", 3600)
                        )
        return token

    async def get_channel_id(self, token: str) -> str:
        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{YOUTUBE_API}/channels?part=id&mine=true",
                headers={"Authorization": f"Bearer {token}"},
            )
            resp.raise_for_status()
            items = resp.json().get("items", [])
            return items[0]["id"] if items else ""

File: src/services/platforms/youtube.ts
import type { PlatformClient, PostPayload, PostResult, Metrics } from "./types";

export class YouTubeClient implements PlatformClient {
  constructor(
    private readonly accessToken: string,
    private readonly refreshToken: string,
    private readonly clientId: string,
    private readonly clientSecret: string,
  ) {}

  private async refreshAccessToken(): Promise<string> {
    const res = await fetch("https://oauth2.googleapis.com/token", {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: this.refreshToken,
        client_id: this.clientId,
        client_secret: this.clientSecret,
      }),
    });
    if (!res.ok) throw new Error("YouTube token refresh failed");
    const data = await res.json() as { access_token: string };
    return data.access_token;
  }

  async post(payload: PostPayload): Promise<PostResult> {
    const token = await this.refreshAccessToken();
    const videoUrl = payload.mediaUrls?.[0];
    if (!videoUrl) throw new Error("YouTube post requires a video URL");

    const metadata = {
      snippet: {
        title: payload.videoTitle ?? payload.content.slice(0, 100),
        description: payload.videoDescription ?? payload.content,
        tags: payload.videoTags ?? [],
      },
      status: { privacyStatus: "public" },
    };

    const initRes = await fetch(
      "https://www.googleapis.com/upload/youtube/v3/videos?uploadType=resumable&part=snippet,status",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
          "X-Upload-Content-Type": "video/*",
        },
        body: JSON.stringify(metadata),
      },
    );
    if (!initRes.ok) throw new Error(`YouTube upload init failed: ${initRes.status}`);
    const uploadUrl = initRes.headers.get("location");
    if (!uploadUrl) throw new Error("No upload URL from YouTube");

    const videoRes = await fetch(videoUrl);
    const videoBlob = await videoRes.blob();
    const uploadRes = await fetch(uploadUrl, {
      method: "PUT",
      headers: { "Content-Type": "video/*" },
      body: videoBlob,
    });
    if (!uploadRes.ok) throw new Error(`YouTube upload failed: ${uploadRes.status}`);
    const data = await uploadRes.json() as { id: string };
    return { platformPostId: data.id };
  }

  async getMetrics(platformPostId: string): Promise<Metrics> {
    const token = await this.refreshAccessToken();
    const res = await fetch(
      `https://www.googleapis.com/youtube/v3/videos?part=statistics&id=${platformPostId}`,
      { headers: { Authorization: `Bearer ${token}` } },
    );
    if (!res.ok) return { likes: 0, shares: 0, comments: 0, views: 0, clicks: 0, rawData: {} };
    const data = await res.json() as { items: Array<{ statistics: Record<string, string> }> };
    const stats = data.items[0]?.statistics ?? {};
    return {
      likes: Number(stats.likeCount ?? 0),
      shares: 0,
      comments: Number(stats.commentCount ?? 0),
      views: Number(stats.viewCount ?? 0),
      clicks: 0,
      rawData: stats as Record<string, unknown>,
    };
  }
}
